src/DeviceInterface.cpp

```diff
@@ -80,7 +80,13 @@ int halide_copy_to_device(void *user_context, struct buffer_t *buf,
 /** Wait for current GPU operations to complete. Calling this explicitly
  * should rarely be necessary, except maybe for profiling. */
 int halide_device_sync(void *user_context, struct buffer_t *buf) {
-    return false;
+    user_assert(user_context == NULL) << "Cannot provide user_context to libHalide.a halide_device_sync\n";
```

Comment:
For JIT, if the user_context is non-null, it is a structure that holds the function pointers that are the callbacks for the JIT realization that made the call into the runtime. The JIT sets its own custom handlers a number of routines, such as halide_print, etc. These handlers look at the user_context and if it is non-null, indirect through it to call the desired target routine. This allows making the set_custom_* methods on Func work while sharing a single copy of the Halide runtime across JIT invocations of different Funcs.  If the user_context is nullptr, then the default runtime routines are used.

(Lower-level, the runtime code contains globals that hold the custom handers. The JIT takes the one global that exists and sets it to a routine that indirects through a table of function pointers, or calls the default routine if the pointer to the table is nullptr. The user_context mechanism is used to pass the table pointer. The table of function pointers has one pointer per type of handler.)

The wrappers in DeviceInterface.cpp are called by the client code and the client code has no way to produce the callback structure, nor should is assume details of the JIT handlers. But if the handlers are passed a non-null user_context, the JIT installed handlers would crash trying to indirect through it.

In effect, user_context is not usable for client code in JIT.

I hope we'll someday replace user_context with something cleaner and more sensible, but for the now, these are the facts of life.